app-misc/uptimed/uptimed-0.4.6-r1.ebuild

```diff
@@ -0,0 +1,45 @@
+# Copyright 2021-2022 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit autotools systemd
+
+DESCRIPTION="System uptime record daemon that keeps track of your highest uptimes"
+HOMEPAGE="https://github.com/rpodgorny/uptimed/"
+SRC_URI="https://github.com/rpodgorny/uptimed/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~mips ~ppc ~ppc64 ~riscv ~sparc ~x86"
+
+RDEPEND="
+	acct-group/uptimed
+	acct-user/uptimed
+"
+DEPEND="${RDEPEND}"
+BDEPEND="${RDEPEND}"
+
+src_prepare() {
+	default
+	eautoreconf
+}
+
+src_install() {
+	local DOCS=( ChangeLog README.md TODO AUTHORS CREDITS INSTALL.cgi sample-cgi/* )
+	default
+	find "${ED}" -type f -name '*.la' -delete || die
+
+	local spooldir="/var/spool/${PN}"
```

Comment:
Your changes looks good. While we are chaing, doesn't make it sense to move ```/var/spool/uptimed``` into init script, which checks and sets correct permissions with ```checkpath```?

Reply:
`/var/spool/uptimed` is persistent (uptimed retains previous uptime data across reboots) so we should only need to set the correct permissions once, at emerge time. I think the old "fixing permissions" stuff was there to address a bug in an ancient version of the ebuild that no longer exists.

Reply:
I can't really follow that logic, since it's very common to create directories (or check) needed for runtime by the init scripts. Sure, once created, it's there. But if we can drop it all, also fine for me.

Reply:
Yeah, I don't think the init script should be modifying permissions. That tends to lead to the same class of vulnerability via the init script rather than `pkg_postinst`, as with [sguil-sensor for example](https://bugs.gentoo.org/603272).

Reply:
To elaborate a bit, it's generally unsafe to fix any permissions as root if you're doing it in a directory that's not owned and writable only by root. So for example, it would be unsafe to adjust the permissions of files in `/var/spool/uptimed`, because that directory is owned by the `uptimed` user (and not root). Even checkpath is not truly safe in that regard because it contains a race condition in the "is this a link?" check.

In contrast, it's OK to create directories directly under `/run` or to tweak their permissions because `/run` is only writable by root. Maybe more to the point, service scripts _have_ to create those directories under `/run`, because `/run` gets wiped on a reboot. So whatever small risks we incur within `/run` can be justified by how hard they are to avoid. But since we _don't_ have to recreate `/var/spool/uptimed`, it's easier and safer to not get involved in the whole mess.

Reply:
> That tends to lead to the same class of vulnerability via the init script rather than pkg_postinst, as with [sguil-sensor for example](https://bugs.gentoo.org/603272).

I didn't meant usage of chown in scripts. I agree here. I meant checkpath/tmpfiles.

> To elaborate a bit, it's generally unsafe to fix any permissions as root if you're doing it in a directory that's not owned and writable only by root. So for example, it would be unsafe to adjust the permissions of files in /var/spool/uptimed, because that directory is owned by the uptimed user (and not root).

I agree here. But I thought, we are talking about directory, not files? 

> Even checkpath is not truly safe in that regard because it contains a race condition in the "is this a link?" check.

Okay, and what's the solution for that?

> But since we don't have to recreate /var/spool/uptimed, it's easier and safer to not get involved in the whole mess.

Well, but it then stills need to be present in ebuild, for first installation variants, where it's still not created, if it would be removed from ebuild then.

Reply:
> 
> I agree here. But I thought, we are talking about directory, not files?
> 

It would be safe to create (or modify) `/var/spool/uptimed` itself, but not the files it contains. The snippet that @ajakk removed,

```bash
        find ${spooldir} -type f -links 1 \
            \( -name records -o -name records.old \) \
            | xargs --no-run-if-empty chown uptimed:uptimed || die
```

affects the files within `/var/spool/uptimed` and there's really no safe way to do that.


> > Even checkpath is not truly safe in that regard because it contains a race condition in the "is this a link?" check.
> 
> Okay, and what's the solution for that?

Essentially, "don't do that." Under `/run` there's not much reason to alter existing permissions because it gets wiped on reboot. But in general it's not safe to do _anything_ as root inside of a directory you don't control. If at all possible, "fixing" permissions and ownership should be avoided in favor of creating them with the correct bits in the first place. For example that might mean switching to the `uptimed` user and setting an appropriate umask before creating stuff, rather than creating it all as root and using chown/chmod to set the owner/permissions.

It would be nice if something like `su` were standardized for the purpose of "do this thing as this user," but all of the utilities that I know of have cross-platform quirks.

> > But since we don't have to recreate /var/spool/uptimed, it's easier and safer to not get involved in the whole mess.
> 
> Well, but it then stills need to be present in ebuild, for first installation variants, where it's still not created, if it would be removed from ebuild then.

Ebuilds install to an empty temporary image that is only accessible by root, and then copy everything over to the live filesystem. If you call chown etc. within the temporary image, it's safe, since any potential attackers can't access the image to mess with things.

There _is_ something of a grey area when it comes to the live filesystem merge, but I haven't investigated it in depth.

Reply:
> It would be safe to create (or modify) /var/spool/uptimed itself, but not the files it contains. The snippet that @ajakk removed,

Eh, I wasn't talking about the removed lines. That's fine. I was talking about removing line 33-35, which only sets permission on the directory, but not files?

Reply:
> Eh, I wasn't talking about the removed lines. That's fine. I was talking about removing line 33-35, which only sets permission on the directory, but not files?

Modulo the keepdir, that'd be safe to do either way. It's just a waste of cycles to check on it each time uptimed starts.